### ISSUE TITLE
Non-determinism over different batch sizes

### ISSUE THREAD

**Issue by JackTemaki** [rwth-i6/returnn#185]
For the following code, the results for a changed output of 'style_weights' depending on the other sequences in the batch:
```
  'pre_slice': {'class': 'slice', 'axis': 'F', 'slice_start': 32, 'from': ['data']},

  'style_conv_layer_1': {'class': 'conv', 'activation': 'relu', 'filter_size':(3,3), 'padding': 'same', 'strides': (2,2), 'n_out': 32, 'from': ['pre_slice'], 'input_add_feature_dim': True},
  'style_conv_layer_2': {'class': 'conv', 'activation': 'relu', 'filter_size':(3,3), 'padding': 'same', 'strides': (2,2), 'n_out': 32, 'from': ['style_conv_layer_1'], 'dropout': refenc_conv_dropout},
  'style_conv_layer_3': {'class': 'conv', 'activation': 'relu', 'filter_size':(3,3), 'padding': 'same', 'strides': (2,2), 'n_out': 64, 'from': ['style_conv_layer_2'], 'dropout': refenc_conv_dropout},
  'style_conv_layer_4': {'class': 'conv', 'activation': 'relu', 'filter_size':(3,3), 'padding': 'same', 'strides': (2,2), 'n_out': 64, 'from': ['style_conv_layer_3'], 'dropout': refenc_conv_dropout},
  'style_conv_layer_5': {'class': 'conv', 'activation': 'relu', 'filter_size':(3,3), 'padding': 'same', 'strides': (2,2), 'n_out': 128, 'from': ['style_conv_layer_4'], 'dropout': refenc_conv_dropout},
  'style_conv_layer_6': {'class': 'conv', 'activation': 'relu', 'filter_size':(3,3), 'padding': 'same', 'strides': (2,2), 'n_out': 128, 'from': ['style_conv_layer_5'], 'dropout': refenc_conv_dropout},

  'reshape_enc': {'class': 'merge_dims', 'from': ['style_conv_layer_6'], 'axes': ['F', 's:-1'], 'n_out': 256},

  'ref_enc': {'class': 'rec', 'direction': 1, 'from': ['reshape_enc'], 'n_out': 128, 'unit': 'nativelstm'},

  'style_tokens': {'class': 'variable', 'shape': ((num_style_tokens, style_token_dim)),'time_axis': 1, 'init': 'glorot_normal', 'out_type': {'dim': style_token_dim, 'shape': (num_style_tokens, style_token_dim), 'time_dim_axis': 1}},
  'style_trans': {'class': 'linear', 'activation': None, 'from': ['style_tokens'], 'n_out': 128},
  'ref_reduce': {'class': 'get_last_hidden_state', 'from': ['ref_enc'], 'combine': 'concat', 'n_out': 128},
  #'ref_reduce': {'class': 'reduce', 'mode': 'max', 'from': ['pre_slice'], 'axes': 'spatial', 'use_time_mask': True}, # this alternative produces the same error
  'ref_trans': {'class': 'linear', 'activation': None, 'from': ['ref_reduce'], 'n_out': 128, 'dropout': 0.2},

  'style_energy_in': {'class': 'combine', 'from': ['style_trans', 'ref_trans'], 'kind': 'add', 'n_out': 128, 'out_type': {'dim': 128, 'shape': (num_style_tokens, 128)}} ,
  'style_tanh': {'activation': 'tanh', 'class': 'activation', 'from': ['style_energy_in'], 'dropout': 0.1},
  'style_energy': {'activation': None, 'class': 'linear', 'from': ['style_tanh'], 'n_out': 1, 'with_bias': False, 'out_type': {'dim': 1, 'shape': (num_style_tokens, 1)}},
  'style_weights': {'class': 'softmax_over_spatial', 'from': ['style_energy']},
```

So for a batch containing sequences 1 and 2 the output is A and B, and for the batch containing the sequences 1 and 3 the output is C and D, so the output for the same sequence changes. My assumption is, that merge_dims does remove the masking of the sequence, but i am not sure. When conditioning 'ref_reduce' directly on the 'pre_slice', this error does not occur.

**Comment by albertz** [comment]
Can you try to come up with a simple test case (for `test_TFNetworkLayer.py`)?
In the test case:
Construct the net, feed in seqs 1 and 2, check output, feed in seqs 1 and 3, check output again, and check that it is the same. This should fail, as you report.
Then try to reduce this test case as far as possible such that you still get this failing.
Then we can debug this. You might be right with your assumption that the problem is in merge_dims. We can extend the test by also checking for that.
Then we should fix the problem.
Then, the test case, together with the fix, would make a nice PR. :)

**Comment by albertz** [comment]
(Btw, off topic to your issue: For style_tokens, you need only single brackets for shape, not `((`. Also, why do you specify out_type? You should not need that. Except that it would set time_dim_axis=None I think. Which would be right, or not? I guess you do that because of the linear transformation then. Maybe better would be to use reinterpret_data explicitly to set the time dim such that the linear layer works as intended.)

**Comment by albertz** [comment]
You say that ref_reduce conditioned directly on pre_slice does not have the same error? But in the code, you write: "this alternative produces the same error". Typo?

**Comment by JackTemaki** [comment]
I figured out a way to use the reinterpret_data layer instead of my custom parameter in the variable layer. Maybe we can discuss separately if it would be nice the variable layer to be able to directly define its axis instead of relying on another layer. For the ref_reduce, this was a typo, yes. So here is the corrected code, where i used max for the working and the non working case:
```
  'pre_slice': {'class': 'slice', 'axis': 'F', 'slice_start': 80, 'from': ['data']},

  'style_conv_layer_1': {'class': 'conv', 'activation': 'relu', 'filter_size':(3,3), 'padding': 'same', 'strides': (2,2), 'n_out': 32, 'from': ['pre_slice'], 'input_add_feature_dim': True},
  'style_conv_layer_2': {'class': 'conv', 'activation': 'relu', 'filter_size':(3,3), 'padding': 'same', 'strides': (2,2), 'n_out': 32, 'from': ['style_conv_layer_1'], 'dropout': refenc_conv_dropout},
  'style_conv_layer_3': {'class': 'conv', 'activation': 'relu', 'filter_size':(3,3), 'padding': 'same', 'strides': (2,2), 'n_out': 64, 'from': ['style_conv_layer_2'], 'dropout': refenc_conv_dropout},
  'style_conv_layer_4': {'class': 'conv', 'activation': 'relu', 'filter_size':(3,3), 'padding': 'same', 'strides': (2,2), 'n_out': 64, 'from': ['style_conv_layer_3'], 'dropout': refenc_conv_dropout},
  'style_conv_layer_5': {'class': 'conv', 'activation': 'relu', 'filter_size':(3,3), 'padding': 'same', 'strides': (2,2), 'n_out': 128, 'from': ['style_conv_layer_4'], 'dropout': refenc_conv_dropout},
  'style_conv_layer_6': {'class': 'conv', 'activation': 'relu', 'filter_size':(3,3), 'padding': 'same', 'strides': (2,2), 'n_out': 128, 'from': ['style_conv_layer_5'], 'dropout': refenc_conv_dropout},

  'reshape_enc': {'class': 'merge_dims', 'from': ['style_conv_layer_6'], 'axes': ['F', 's:-1'], 'n_out': 256},

  'ref_enc': {'class': 'rec', 'direction': 1, 'from': ['reshape_enc'], 'n_out': 128, 'unit': 'nativelstm'},
  'style_tokens_variable': {'class': 'variable', 'shape': (num_style_tokens, style_token_dim), 'init': 'glorot_normal'},
  'style_tokens': {'class': 'reinterpret_data', 'from': ['style_tokens_variable'], 'set_axes': {'T': 'spatial'}},
  'style_trans': {'class': 'linear', 'activation': None, 'from': ['style_tokens'], 'n_out': 128},

  #'ref_reduce': {'class': 'reduce', 'mode': 'max', 'from': ['pre_slice'], 'axes': 'spatial', 'use_time_mask': True}, # this works when setting pre_slice to be 128 dimensional
  'ref_reduce': {'class': 'reduce', 'mode': 'max', 'from': ['ref_enc'], 'axes': 'spatial', 'use_time_mask': True}, # this produces the error

  'ref_trans': {'class': 'linear', 'activation': None, 'from': ['ref_reduce'], 'n_out': 128, 'dropout': 0.2},

  'style_energy_in': {'class': 'combine', 'from': ['style_trans', 'ref_trans'], 'kind': 'add', 'n_out': 128} ,
  'style_tanh': {'activation': 'tanh', 'class': 'activation', 'from': ['style_energy_in'], 'dropout': 0.1},
  'style_energy': {'activation': None, 'class': 'linear', 'from': ['style_tanh'], 'n_out': 1, 'with_bias': False},
  'style_weights': {'class': 'softmax_over_spatial', 'from': ['style_energy']},
```

I will try to formulate this as a test case

**Comment by albertz** [comment]
But I still don't understand: Why are you doing this? Why do you need the time dim there?

Also, pre_slice does not make much sense to me. Why do you do this?

Btw, it would also be helpful if you post also the output via `debug_print_layer_output_template`. I'm not sure I get it all right from just looking at it. (You don't have comments in the code about the shape. I think it is a good habit, in such cases, to add a comment describing the shape you expect.)

The style_weights which does softmax_over_spatial, over what axis exactly do you intend it to operate? As I understand the config, this is definitely not the input time axis.

Btw, you have dropout also. Are you sure that when you checked it, it was deterministic? Maybe disable dropout to make sure.

**Comment by JackTemaki** [comment]
I found out that this is not a bug, but intended behaviour from Tensorflow when using padding="same" and a stride > 1 in convolutions. If this is the case, Tensorflow will set the filter positions based on the maximum length in the batch, and this can lead to "shifted" filter positions by 1 when the batch size changes. As there are 6 Layers in the encoder, usually at least one layer will have this shift when changing the batch size. Possible fixes are using stride 1 and max-pooling or padding="valid". While testing I also found out that batch-norm has slight non-determinism with regards to the batch size.
